runtime/vm/ContinuationHelpers.cpp

```diff
@@ -141,19 +176,23 @@ yieldContinuation(J9VMThread *currentThread)
 {
 	BOOLEAN result = TRUE;
 	J9VMContinuation *continuation = currentThread->currentContinuation;
-
 	Assert_VM_notNull(currentThread->currentContinuation);
 
-	VM_ContinuationHelpers::swapFieldsWithContinuation(currentThread, continuation);
-	continuation->carrierThread = NULL;
+	/* we don't need atomic here, since no GC thread should be able to start scanning while continuation is mounted,
+	 * nor should another carrier thread be able to mount before we complete the unmount (hence no risk to overwrite anything in a race).
+	 */
+	Assert_VM_false(VM_VMHelpers::isConcurrentlyScannedFromContinuationState(continuation->state));
```

Comment:
I understand your intention to put this comment early, thinking if there is a need for synchronization, it should be done early.

But comment really applies to putting state back to initial, while someone may assume we are talking about currentContinuation being set to NULL. So either we should state it more explicitly rather than saying 'here" or (what was my initial intention) just keep the comment next to state reset.

Same applies to the assert, it's the best to be the closest to state reset as possible (in some malicious scenario if anotehr thread mutates state between the assert and the reset, keeping them close increases chances to catch the intruder)
Then we can merge the 2 asserts into one: currentThread == state. Such an assert would not be completely clean (in a sense it knows something about how state struct is organized), but I'm still ok with it, for the sake of simpler code (asserts). Besides, it would not be the only spot: for example, atomic in tryWinningConcurrentGCScan also assumes this knowledge.

Reply:
we'll need a comment about the reset being the last step and write fence preceding it, but only after/if GAC confirms 'borrowing' the existing fence was ok, first